ReadWriteOncePod access mode alpha feature blog (#29059)
kubernetes/website: content/en/blog/_posts/2021-07-16-read-write-once-pod-access-mode-alpha.md

---
layout: blog
title: "Introducing Single Pod Access Mode for PersistentVolumes"
date: 2021-07-16
slug: read-write-once-pod-access-mode-alpha
---

**Author:** Chris Henzie (Google)

Kubernetes 1.22 introduces a new ReadWriteOncePod access mode for PersistentVolumes and PersistentVolumeClaims that allows users to restrict volume access to a single pod in the cluster.

## What are access modes and why are they important?

When using storage, there are different ways we can model how it is consumed.

For example, a storage system like a network file share can have many users all reading and writing data simultaneously.
In other cases maybe everyone is allowed to read data but not write it.
For highly sensitive data, maybe only one user is allowed to read and write data but nobody else.

In the world of Kubernetes, access modes are the way we model how storage is consumed.
These access modes are a part of the spec for PersistentVolumes (PVs) and PersistentVolumeClaims (PVCs).

> <!-- TODO: Is there a way to strongly highlight the accessModes section inside this block? -->

Reviewer: We usually use YAML comments for that kind of highlighting.

metadata:
  name: cat-pictures
spec:
  accessModes:
  - ReadWriteMany
  resources:
    requests:
      storage: 1Gi
```

In Kubernetes, we currently have three access modes for PVs and PVCs:

- ReadWriteOnce -- the volume can be mounted as read-write by a single node
- ReadOnlyMany -- the volume can be mounted read-only by many nodes
- ReadWriteMany -- the volume can be mounted as read-write by many nodes

These access modes are enforced by Kubernetes components like the kube-controller-manager and kubelet to ensure only certain pods are allowed to access a given PersistentVolume.
This allows users to follow the principle of least privilege to mitigate the impact of bad actors.

## What is this new access mode and how does it work?

In Kubernetes 1.22, a fourth access mode for PVs and PVCs is introduced:

- ReadWriteOncePod -- the volume can be mounted as read-write by a single pod

When a user creates a pod with a PVC that has the ReadWriteOncePod access mode, Kubernetes will ensure that pod is the only pod in the cluster that can read and write to it.

If you create another pod that references the same PVC with this access mode, the pod will fail to start like so:

> ```shell

Reviewer: This isn't a shell script. Best not to set any syntax highlighting.

Events:
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  1s    default-scheduler  0/1 nodes are available: 1 node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode.
```

### How is this different than the ReadWriteOnce access mode?

> The ReadWriteOnce access mode restricts volume access to a single *node*, which means it is possible for multiple pods on the same node to write to the same volume.

Reviewer: Is this bad (potentially yes!) - let's explain the benefit. Maybe [fencing](https://en.wikipedia.org/wiki/Fencing_(computing)) a legacy app?

## How do I use it?

> The ReadWriteOncePod access mode is in alpha for Kubernetes 1.22. As a first step we need to enable the ReadWriteOncePod feature gate for `kube-apiserver`, `kube-scheduler`, and `kubelet`. The feature gate can be enabled or disabled as follows (as described in more detail [here]).

Reviewer: We'd typically write:
```suggestion
The ReadWriteOncePod access mode is in alpha for Kubernetes 1.22. As a first step you need to enable the ReadWriteOncePod [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates) for `kube-apiserver`, `kube-scheduler`, and `kubelet`. You can enable (or disable) the feature by setting command line arguments:
```

```shell
--feature-gates="...,ReadWriteOncePod=true"
```

[here]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

### Creating a PersistentVolumeClaim

In order to use the ReadWriteOncePod access mode for your PVs and PVCs, you will need to create a new PVC with the access mode:

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: top-secret
spec:
  accessModes:
  - ReadWriteOncePod
  resources:
    requests:
      storage: 1Gi
```

## What volume plugins support this?

## As a storage vendor, how do I add support for this access mode to my CSI driver?

> ### New Controller and Node Service Capabilities

Reviewer: ```suggestion
### New controller and Node Service capabilities
```
Is that capitalization right? We usually put headings in sentence case.

Author: "Node Service" here refers to the [container storage interface](https://github.com/container-storage-interface/spec/blob/master/spec.md#rpc-interface), and not the Kubernetes primitives; removing the capitalization

---
layout: blog
title: "Introducing Single Pod Access Mode for PersistentVolumes"
date: 2021-07-16
slug: read-write-once-pod-access-mode-alpha
---

**Author:** Chris Henzie (Google)

> Kubernetes 1.22 introduces a new ReadWriteOncePod access mode for PersistentVolumes and PersistentVolumeClaims that allows users to restrict volume access to a single pod in the cluster.

Reviewer: Optionally: link PersistentVolumes and PersistentVolumeClaims to https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistent-volumes and https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims (respectively)

## What are access modes and why are they important?

> When using storage, there are different ways we can model how it is consumed.

Reviewer: I'd write:
```suggestion
When using storage, there are different ways to model how that storage is consumed.
```

How does that sound?

Author: Sounds good to me!

For example, a storage system like a network file share can have many users all reading and writing data simultaneously.
In other cases maybe everyone is allowed to read data but not write it.
For highly sensitive data, maybe only one user is allowed to read and write data but nobody else.

> In the world of Kubernetes, [access modes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes) are the way we model how storage is consumed.

Reviewer: ```suggestion
In the world of Kubernetes, [access modes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes) are the way you can define how durable storage is consumed.
```
?

Author: Thanks!

These access modes are a part of the spec for PersistentVolumes (PVs) and PersistentVolumeClaims (PVCs).

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: shared-cache
spec:
  accessModes:
  - ReadWriteMany # Allow many pods to access shared-cache simultaneously.
  resources:
    requests:
      storage: 1Gi
  storageClassName: ""
```

> In Kubernetes, we currently have three access modes for PVs and PVCs:

Reviewer: ```suggestion
Before v1.22, Kubernetes offered three access modes for PVs and PVCs:
```

Reviewer: As this article will be published when **v1.22** is the current release, we really ought to fix the tense of the verb / sense of the text here. What's drafted here would be factually inaccurate at the time it is published, and might also mislead some readers until they read on.

Author: Oops, I included this change but hadn't pushed it. Updated to use the suggested text

- ReadWriteOnce -- the volume can be mounted as read-write by a single node
- ReadOnlyMany -- the volume can be mounted read-only by many nodes

> - ReadWriteMany -- the volume can be mounted as read-write by many nodes

Reviewer: nit: I would use an en dash here (`&ndash;`), rather than `--` (check the [preview](https://deploy-preview-29059--kubernetes-io-main-staging.netlify.app/blog/2021/07/16/read-write-once-pod-access-mode-alpha/) to see how these render)

These access modes are enforced by Kubernetes components like the `kube-controller-manager` and `kubelet` to ensure only certain pods are allowed to access a given PersistentVolume.

## What is this new access mode and how does it work?

> In Kubernetes v1.22, a fourth access mode for PVs and PVCs is introduced for CSI volumes:

Reviewer: ```suggestion
Kubernetes v1.22 introduced a fourth access mode for PVs and PVCs, that you can use for CSI volumes:
```

> - ReadWriteOncePod -- the volume can be mounted as read-write by a single pod

Reviewer: nit: I would use an en dash here (`&ndash;`) too.

> When a user creates a pod with a PVC that has the ReadWriteOncePod access mode, Kubernetes will ensure that pod is the only pod in the cluster that can read and write to it.

Reviewer: I couldn't work out from the existing wording whether ReadWriteOncePod blocks another Pod having just read access to that nonvolatile storage.
```suggestion
If you create a pod with a PVC that uses the ReadWriteOncePod access mode, Kubernetes ensures that new pod is the only pod across your whole the cluster that can read that PVC or write to it.
```

Author: Thank you for the suggestion. The tricky thing about this wording is "new", it implies a new pod will usurp control over the PVC if it's already in-use by another pod. I kept the sentence but removed the word "new", does it make sense without it?

> If you create another pod that references the same PVC with this access mode, the pod will fail to start like so:

Reviewer: ```suggestion
If you create another pod that references the same PVC then, no matter what access mode
you suggest for the new pod, the ReadWriteOncePod access mode on the existing pod
means that the new pod fails to start. For example:
```

Author: These access modes are not on pods, but on the PVC. I provided an explanation for why it won't start. How does this look?

```
Events:
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  1s    default-scheduler  0/1 nodes are available: 1 node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode.
```

### How is this different than the ReadWriteOnce access mode?

The ReadWriteOnce access mode restricts volume access to a single *node*, which means it is possible for multiple pods on the same node to read from and write to the same volume.
This could potentially be a major problem for some applications, especially if they require at most one writer for data safety guarantees.

> With ReadWriteOncePod these issues go away.
> Just set the access mode on your PVC, and Kubernetes will guarantee only a single pod has access.

Reviewer: For technical writing, I try to avoid “just” in this sense - and especially when there _are_ extra steps, such as enabling feature gates across multiple components.

```suggestion
With ReadWriteOncePod these issues go away:
set the access mode on your PVC, and Kubernetes guarantees that only a single pod has access.
```

```shell
kubectl patch pv cat-pictures-pv -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
```

Next you need to stop any workloads that are using the PersistentVolumeClaim bound to the PersistentVolume you want to migrate, and then delete the PersistentVolumeClaim:

```shell
kubectl scale --replicas=0 deployment cat-pictures-writer
kubectl delete pvc cat-pictures-pvc
```

After that you need to replace the PersistentVolume's access modes with ReadWriteOncePod:

```shell
kubectl patch pv cat-pictures-pv -p '{"spec":{"accessModes":["ReadWriteOncePod"]}}'
```

{{< note >}}
The ReadWriteOncePod access mode cannot be combined with other access modes.
Make sure ReadWriteOncePod is the only access mode on the PersistentVolume when updating, otherwise the request will fail.
{{< /note >}}

You can now recreate your PersistentVolumeClaim and start up your workloads:

```shell
kubectl apply -f cat-pictures-pvc.yaml
kubectl apply -f cat-pictures-writer-deployment.yaml
```

Lastly you may edit your PersistentVolume's `spec.persistentVolumeReclaimPolicy` and set to it back to `Delete` if you previously changed it.

```
kubectl edit pv cat-pictures-pv
```

> [Please see this guide](https://kubernetes.io/docs/tasks/configure-pod-container/configure-persistent-volume-storage/) for more details on working with PersistentVolumes and PersistentVolumeClaims.

Reviewer: ```suggestion
You can read [Configure a Pod to Use a PersistentVolume for Storage](/docs/tasks/configure-pod-container/configure-persistent-volume-storage/) for more details on working with PersistentVolumes and PersistentVolumeClaims.
```

## What volume plugins support this?

The only volume plugins that support this are CSI drivers.

> We do not plan to support this for in-tree plugins because they are being deprecated as part of [CSI migration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/#what-is-the-timeline-status).

Reviewer: Readers might not know if “we” refers to Google, the Kubernetes project, or SIG Storage. Can you clarify?

Author: Clarified to SIG Storage

## As a storage vendor, how do I add support for this access mode to my CSI driver?

The ReadWriteOncePod access mode will work out of the box without any required updates to CSI drivers.
With that being said, if you would like to stay up to date with the latest changes to the CSI specification (v1.5.0+), read on.

Two new access modes were introduced to the CSI specification in order to disambiguate the legacy [`SINGLE_NODE_WRITER`](https://github.com/container-storage-interface/spec/blob/v1.5.0/csi.proto#L418-L420) access mode.
They are [`SINGLE_NODE_SINGLE_WRITER` and `SINGLE_NODE_MULTI_WRITER`](https://github.com/container-storage-interface/spec/blob/v1.5.0/csi.proto#L437-L447).
In order to communicate to sidecars (like the [external-provisioner](https://github.com/kubernetes-csi/external-provisioner)) that your driver understands and accepts these two new CSI access modes, your driver will also need to advertise the `SINGLE_NODE_MULTI_WRITER` capability for the [controller service](https://github.com/container-storage-interface/spec/blob/v1.5.0/csi.proto#L1073-L1081) and [node service](https://github.com/container-storage-interface/spec/blob/v1.5.0/csi.proto#L1515-L1524).

> [Please see the KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md#csi-specification-changes-volume-capabilities) for more context on the motivation for these access modes and capability bits.

Reviewer: ```suggestion
If you'd like to read up on the motivation for these access modes and capability bits, you can also read
the [CSI Specification Changes, Volume Capabilities](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md#csi-specification-changes-volume-capabilities)
section of KEP-2485 (ReadWriteOncePod PersistentVolume Access Mode).
```

> ### Update your CSI spec

Reviewer: ```suggestion
### Update your CSI driver to use the new interface
```

```

### Implement `NodePublishVolume` behavior

The CSI spec outlines expected behavior for the `NodePublishVolume` RPC when called more than once for the same volume but with different arguments (like the target path).
Please refer to [the second table in the NodePublishVolume section of the CSI spec](https://github.com/container-storage-interface/spec/blob/v1.5.0/spec.md#nodepublishvolume) for more details on expected behavior when implementing in your driver.

### Update your CSI sidecars (optional)

When deploying your CSI drivers, you may update your CSI sidecars (mainly the [external-provisioner](https://github.com/kubernetes-csi/external-provisioner), [external-attacher](https://github.com/kubernetes-csi/external-attacher), and [external-resizer](https://github.com/kubernetes-csi/external-resizer)) to versions that depend on CSI spec v1.5.0+ and the Kubernetes v1.22 API.

Note that this is not required, as the ReadWriteOncePod access mode will work out of the box without any changes to your driver.
However, if you would like to leverage the new CSI access modes and capabilities added to your driver, this step is recommended.

## What’s next?

> As part of the beta graduation for this feature, support will be added for pod preemption in the scheduler.

Reviewer: - I'd write “we plan to” for statements like this (exception: SIG Architecture have signed off on a commitment that it will happen this way)
- the scheduler already supports preemption; I've reworded to clarify what would be added
```suggestion
As part of the beta graduation for this feature, we plan to update the Kubenetes scheduler to support pod preemption in relation to ReadWriteOncePod storage.
```

(also see separate comment about making sure the reader is clear who “we” refers to).

```shell
kubectl patch pv cat-pictures-pv -p '{"spec":{"persistentVolumeReclaimPolicy":"Retain"}}'
```

Next you need to stop any workloads that are using the PersistentVolumeClaim bound to the PersistentVolume you want to migrate, and then delete the PersistentVolumeClaim:

```shell
kubectl scale --replicas=0 deployment cat-pictures-writer
kubectl delete pvc cat-pictures-pvc
```

After that you need to replace the PersistentVolume's access modes with ReadWriteOncePod:

```shell
kubectl patch pv cat-pictures-pv -p '{"spec":{"accessModes":["ReadWriteOncePod"]}}'
```

{{< note >}}
The ReadWriteOncePod access mode cannot be combined with other access modes.
Make sure ReadWriteOncePod is the only access mode on the PersistentVolume when updating, otherwise the request will fail.
{{< /note >}}

You can now recreate your PersistentVolumeClaim and start up your workloads:

```shell
kubectl apply -f cat-pictures-pvc.yaml
kubectl apply -f cat-pictures-writer-deployment.yaml
```

Lastly you may edit your PersistentVolume's `spec.persistentVolumeReclaimPolicy` and set to it back to `Delete` if you previously changed it.

```

> kubectl edit pv cat-pictures-pv

Reviewer 2: use the patch command again

```

You can read [Configure a Pod to Use a PersistentVolume for Storage](/docs/tasks/configure-pod-container/configure-persistent-volume-storage/) for more details on working with PersistentVolumes and PersistentVolumeClaims.

## What volume plugins support this?

The only volume plugins that support this are CSI drivers.

> SIG Storage does not plan to support this for in-tree plugins because they are being deprecated as part of [CSI migration](/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/#what-is-the-timeline-status).

Reviewer 2: Do we want to modify this to say something like "we may consider support for beta"

Author: Added:
"Support may be considered for beta for users that prefer to use the legacy in-tree volume APIs with CSI migration enabled."

## What is this new access mode and how does it work?

In Kubernetes v1.22, a fourth access mode for PVs and PVCs is introduced:

- ReadWriteOncePod -- the volume can be mounted as read-write by a single pod

When a user creates a pod with a PVC that has the ReadWriteOncePod access mode, Kubernetes will ensure that pod is the only pod in the cluster that can read and write to it.

If you create another pod that references the same PVC with this access mode, the pod will fail to start like so:

```
Events:
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  1s    default-scheduler  0/1 nodes are available: 1 node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode.
```

### How is this different than the ReadWriteOnce access mode?

The ReadWriteOnce access mode restricts volume access to a single *node*, which means it is possible for multiple pods on the same node to read from and write to the same volume.

Could this cause problems?
Potentially yes!

> If you are running workloads that access sensitive data, other pods on that node can access it (if updated to mount the volume).

Reviewer 2: I would not put security as the reason for this feature. We have other controls (PSP, PV being cluster scoped) to prevent pods from other namespaces to access the volume. Using RWOP for security is timing sensitive too. If the other workload gets scheduled first it will get access to the volume.

The primary reason I see is to provide data safety guarantees to applications that can only have at-most one writer.

Author: Thank you for the clarification!

Unfortunately, the only way to work around this is by running a single pod on the node, which is an inefficient use of compute resources in your cluster.

With ReadWriteOncePod these restrictions go away.
You don't have to worry about neighboring pods potentially accessing your volumes.
Just set the access mode on your PVC, and Kubernetes will guarantee only a single pod has access.

> ## How do I use it?

Reviewer 2: Note that this is only supported for CSI plugins.

As an aside, what happens if you enable this on an intree plugin?

Author: Done. If you enable this on an intree plugin it should work, we just have to add ReadWriteOncePod to [the list of supported access modes](https://github.com/kubernetes/kubernetes/blob/v1.22.0-rc.0/pkg/volume/gcepd/gce_pd.go#L113-L118).

Reviewer 2: What was the reason why we decided not to support it for in-tree volumes if nothing stops a user from using it and it works?

Author: See https://github.com/kubernetes/website/pull/29059#issuecomment-889289648

Author: If we wanted to open this up to in-tree drivers we could do this for beta graduation.

Author: Synced with @Jiawei0227 , I did not know the API fields for in-tree drivers were not being removed. If customers want to rely on the legacy API fields + CSI migration and we feel it will help target more users we can add support here

Reviewer 2: Yeah I think given that the enforcement is all done in Kubernetes, the plugin is not involved, and we don't error/reject if someone uses it with an in-tree type, it is better to support it for in-tree types as well so it's consistent across environments.

Reviewer 3: If the implementation of ReadWriteOncePod is done in kubelet or KCM not in each intree plugins code, I think it is fine to support it for intree as well. 

What I was thinking is if we need to edit intree plugin codes that will be not worth it as most of the PersistentVolume plugin types will be either deprecated or removed.

As a first step you need to enable the ReadWriteOncePod [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates) for `kube-apiserver`, `kube-scheduler`, and `kubelet`.
You can enable (or disable) the feature by setting command line arguments:

```
--feature-gates="...,ReadWriteOncePod=true"
```

[here]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

### Creating a PersistentVolumeClaim

In order to use the ReadWriteOncePod access mode for your PVs and PVCs, you will need to create a new PVC with the access mode:

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: top-secret
spec:
  accessModes:
  - ReadWriteOncePod # Allow only a single pod to access top-secret.
  resources:
    requests:
      storage: 1Gi
```

If your storage plugin supports [dynamic provisioning](https://kubernetes.io/docs/concepts/storage/dynamic-provisioning/), new PersistentVolumes will be created with the ReadWriteOncePod access mode applied.

#### Migrating existing PersistentVolumes

If you have existing PersistentVolumes, they can be migrated to use ReadWriteOncePod.

> As a prerequisite, you need to stop any workloads that are using the PersistentVolumeClaim bound to the PersistentVolume you want to migrate, and then delete the PersistentVolumeClaim:

Reviewer 2: Before you delete the PVC, you must change the PV.spec.reclaimPolicy to Retain, otherwise, the disk will get deleted.

```
kubectl delete pvc cat-pictures-pvc
```

Next you need to replace the PersistentVolume's access modes with ReadWriteOncePod:

```yaml

> kind: PersistentVolume

Reviewer 2: You will probably want to set PVC.storageClassName to "", otherwise it will use the default storageclass.

Or set the PV.StorageClassName to the same storageClassName that the original PVC was.

apiVersion: v1
metadata:
  name: cat-pictures-pv

> spec:

Reviewer 2: You should also set the spec.ClaimRef to the name/namespace of the PVC. That reserves the PV for that PVC and prevents the PV from being claimed by another PVC.

  accessModes:

>     - ReadWriteOnce    # Delete existing access modes like ReadWriteOnce.

Reviewer 2: Maybe RWO should be commented out to facilitate someone copy-pasting the example.

> kubectl patch pv cat-pictures-pv -p '{"spec":{"accessModes":["ReadWriteOncePod"]}}'

Reviewer 2: Oh you may also need to remove the pv.spec.ClaimRef.UID field in order to rebind to the new PVC. You may need to test out this sequence to make sure it works

Author: Thank you for pointing this out, I encountered this issue when running through the steps

                csi.ControllerServiceCapability_RPC_LIST_VOLUMES,
                csi.ControllerServiceCapability_RPC_LIST_VOLUMES_PUBLISHED_NODES,
+               csi.ControllerServiceCapability_RPC_SINGLE_NODE_MULTI_WRITER,
        }
        gceDriver.AddControllerServiceCapabilities(csc)
        ns := []csi.NodeServiceCapability_RPC_Type{
                csi.NodeServiceCapability_RPC_STAGE_UNSTAGE_VOLUME,
                csi.NodeServiceCapability_RPC_EXPAND_VOLUME,
                csi.NodeServiceCapability_RPC_GET_VOLUME_STATS,
+               csi.NodeServiceCapability_RPC_SINGLE_NODE_MULTI_WRITER,
        }
        gceDriver.AddNodeServiceCapabilities(ns)
```

### Implement `NodePublishVolume` behavior

The CSI spec outlines expected behavior for the `NodePublishVolume` RPC when called more than once for the same volume but with different arguments (like the target path).
Please refer to [the second table in the NodePublishVolume section of the CSI spec](https://github.com/container-storage-interface/spec/blob/v1.5.0/spec.md#nodepublishvolume) for more details on expected behavior when implementing in your driver.

### Update your CSI sidecars (optional)

> When deplyoing your CSI drivers, you may update your CSI sidecars (mainly the [external-provisioner](https://github.com/kubernetes-csi/external-provisioner), [external-attacher](https://github.com/kubernetes-csi/external-attacher), and [external-resizer](https://github.com/kubernetes-csi/external-resizer)) to versions that depend on CSI spec v1.5.0+ and the Kubernetes v1.22 API.

Reviewer 2: typo: deploying

## How can I learn more?

Please see [KEP-2485](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md) for more details on the ReadWriteOncePod access mode and motivations for CSI spec changes.

## How do I get involved?

The [Kubernetes #csi Slack channel](https://kubernetes.slack.com/messages/csi) and any of the [standard SIG Storage communication channels](https://github.com/kubernetes/community/blob/master/sig-storage/README.md#contact) are great mediums to reach out to the SIG Storage and the CSI teams.

Special thanks to the following people for their insightful reviews and design considerations:

* Michelle Au (msau42)
* Ben Swartzlander (bswartz)
* Jan Šafránek (jsafrane)
* Hemant Kumar (gnufied)
* Xing Yang (xing-yang)
* Saad Ali (saad-ali)
* Jing Xu (jingxu97)
* Tim Hockin (thockin)
* Jordan Liggitt (liggitt)
* Deep Debroy (ddebroy)
* James DeFelice (jdef)
* Humble Devassy Chirammal (humblec)
* Aldo Culquicondor (alculquicondor)

> * ahg-g

Reviewer 2: Abdullah Gharaibeh, from https://github.com/kubernetes/community/tree/master/sig-scheduling

```

Next you need to stop any workloads that are using the PersistentVolumeClaim bound to the PersistentVolume you want to migrate, and then delete the PersistentVolumeClaim.

Once that is done, you need to clear your PersistentVolume's `spec.claimRef.uid` to ensure PersistentVolumeClaims can bind to it upon recreation:

```shell
kubectl scale --replicas=0 deployment cat-pictures-writer
kubectl delete pvc cat-pictures-pvc
kubectl patch pv cat-pictures-pv -p '{"spec":{"claimRef":{"uid":""}}}'
```

After that you need to replace the PersistentVolume's access modes with ReadWriteOncePod:

```shell
kubectl patch pv cat-pictures-pv -p '{"spec":{"accessModes":["ReadWriteOncePod"]}}'
```

{{< note >}}
The ReadWriteOncePod access mode cannot be combined with other access modes.
Make sure ReadWriteOncePod is the only access mode on the PersistentVolume when updating, otherwise the request will fail.
{{< /note >}}

Next you need to modify your PersistentVolumeClaim to set ReadWriteOncePod as the only access mode.
You should also set your PersistentVolumeClaim's `spec.volumeName` to the name of your PersistentVolume.

> Doing this guarantees your PersistentVolumeClaim will bind to this specific PersistentVolume.

Reviewer 2: It's actually the opposite (setting PV.spec.claimref) that guarantees the binding. If you only set the PVC.spec.volumeName, you could run into an issue where multiple PVCs set the same PV.

---
layout: blog
title: "Introducing Single Pod Access Mode for PersistentVolumes"
date: 2021-07-16
slug: read-write-once-pod-access-mode-alpha
---

**Author:** Chris Henzie (Google)

> Kubernetes v1.22 introduces a new ReadWriteOncePod access mode for [PersistentVolumes](/docs/concepts/storage/persistent-volumes/#persistent-volumes) and [PersistentVolumeClaims](/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims) that allows users to restrict volume access to a single pod in the cluster.

Reviewer: ```suggestion
Last month's release of Kubernetes v1.22 introduced a new ReadWriteOncePod access mode for [PersistentVolumes](/docs/concepts/storage/persistent-volumes/#persistent-volumes) and [PersistentVolumeClaims](/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims). With this alpha feature, Kubernetes allows you to restrict volume access to a single pod in the cluster.
```
_or_
```suggestion
Last month's release of Kubernetes v1.22 introduced a new ReadWriteOncePod access mode for [PersistentVolumes](/docs/concepts/storage/persistent-volumes/#persistent-volumes) and [PersistentVolumeClaims](/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims). With this alpha feature, Kubernetes allows you to define volumes that only a single pod can access at a time.
```

## What are access modes and why are they important?

When using storage, there are different ways to model how that storage is consumed.

For example, a storage system like a network file share can have many users all reading and writing data simultaneously.
In other cases maybe everyone is allowed to read data but not write it.
For highly sensitive data, maybe only one user is allowed to read and write data but nobody else.

In the world of Kubernetes, [access modes](/docs/concepts/storage/persistent-volumes/#access-modes) are the way you can define how durable storage is consumed.
These access modes are a part of the spec for PersistentVolumes (PVs) and PersistentVolumeClaims (PVCs).

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: shared-cache
spec:
  accessModes:
  - ReadWriteMany # Allow many pods to access shared-cache simultaneously.
  resources:
    requests:
      storage: 1Gi

>   storageClassName: ""

Reviewer: Is that empty value right? Could we omit it?

Author: Yes we can, it will default to the customer's default storage class (we do the same thing with the example on like 93).

```

Before v1.22, Kubernetes offered three access modes for PVs and PVCs:

- ReadWriteOnce &ndash; the volume can be mounted as read-write by a single node
- ReadOnlyMany &ndash; the volume can be mounted read-only by many nodes
- ReadWriteMany &ndash; the volume can be mounted as read-write by many nodes

These access modes are enforced by Kubernetes components like the `kube-controller-manager` and `kubelet` to ensure only certain pods are allowed to access a given PersistentVolume.

## What is this new access mode and how does it work?

Kubernetes v1.22 introduced a fourth access mode for PVs and PVCs, that you can use for CSI volumes:

- ReadWriteOncePod &ndash; the volume can be mounted as read-write by a single pod

> If you create a pod with a PVC that uses the ReadWriteOncePod access mode, Kubernetes ensures that pod is the only pod across your whole the cluster that can read that PVC or write to it.

Reviewer 4: ```suggestion
If you create a pod with a PVC that uses the ReadWriteOncePod access mode, Kubernetes ensures that pod is the only pod across your whole cluster that can read that PVC or write to it.
```

Author: Oops, fixed

For example:

```
Events:
  Type     Reason            Age   From               Message
  ----     ------            ----  ----               -------
  Warning  FailedScheduling  1s    default-scheduler  0/1 nodes are available: 1 node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode.
```

### How is this different than the ReadWriteOnce access mode?

The ReadWriteOnce access mode restricts volume access to a single *node*, which means it is possible for multiple pods on the same node to read from and write to the same volume.
This could potentially be a major problem for some applications, especially if they require at most one writer for data safety guarantees.

With ReadWriteOncePod these issues go away.
Set the access mode on your PVC, and Kubernetes guarantees that only a single pod has access.

## How do I use it?

The ReadWriteOncePod access mode is in alpha for Kubernetes v1.22 and is only supported for CSI volumes.
As a first step you need to enable the ReadWriteOncePod [feature gate](/docs/reference/command-line-tools-reference/feature-gates) for `kube-apiserver`, `kube-scheduler`, and `kubelet`.

> You can enable (or disable) the feature by setting command line arguments:

Reviewer 4: ```suggestion
You can enable the feature by setting command line arguments:
```

                csi.ControllerServiceCapability_RPC_LIST_VOLUMES,
                csi.ControllerServiceCapability_RPC_LIST_VOLUMES_PUBLISHED_NODES,
+               csi.ControllerServiceCapability_RPC_SINGLE_NODE_MULTI_WRITER,
        }
        gceDriver.AddControllerServiceCapabilities(csc)
        ns := []csi.NodeServiceCapability_RPC_Type{
                csi.NodeServiceCapability_RPC_STAGE_UNSTAGE_VOLUME,
                csi.NodeServiceCapability_RPC_EXPAND_VOLUME,
                csi.NodeServiceCapability_RPC_GET_VOLUME_STATS,
+               csi.NodeServiceCapability_RPC_SINGLE_NODE_MULTI_WRITER,
        }
        gceDriver.AddNodeServiceCapabilities(ns)
```

### Implement `NodePublishVolume` behavior

The CSI spec outlines expected behavior for the `NodePublishVolume` RPC when called more than once for the same volume but with different arguments (like the target path).
Please refer to [the second table in the NodePublishVolume section of the CSI spec](https://github.com/container-storage-interface/spec/blob/v1.5.0/spec.md#nodepublishvolume) for more details on expected behavior when implementing in your driver.

> ### Update your CSI sidecars (optional)

Reviewer: only a nit, totally fine to leave as-is:
```suggestion
### Update your CSI sidecars (optional) {#update-csi-sidecars}
```

## How can I learn more?

Please see [KEP-2485](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md) for more details on the ReadWriteOncePod access mode and motivations for CSI spec changes.

## How do I get involved?

The [Kubernetes #csi Slack channel](https://kubernetes.slack.com/messages/csi) and any of the [standard SIG Storage communication channels](https://github.com/kubernetes/community/blob/master/sig-storage/README.md#contact) are great mediums to reach out to the SIG Storage and the CSI teams.

Special thanks to the following people for their insightful reviews and design considerations:

* Michelle Au (msau42)
* Ben Swartzlander (bswartz)
* Jan Šafránek (jsafrane)
* Hemant Kumar (gnufied)
* Xing Yang (xing-yang)
* Saad Ali (saad-ali)
* Jing Xu (jingxu97)
* Tim Hockin (thockin)
* Jordan Liggitt (liggitt)
* Deep Debroy (ddebroy)
* James DeFelice (jdef)
* Humble Devassy Chirammal (humblec)
* Aldo Culquicondor (alculquicondor)

> * Abdullah Gharaibeh (ahg-g)

Reviewer: nit: I might consider sorting that list alphabetically.

---
layout: blog
title: "Introducing Single Pod Access Mode for PersistentVolumes"

> date: 2021-07-16

Reviewer 5: This is the date that needs to be updated.

I'd use the date Tim suggested.

## How do I use it?

The ReadWriteOncePod access mode is in alpha for Kubernetes 1.22 and is only supported for CSI volumes.
As a first step you need to enable the ReadWriteOncePod [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates) for `kube-apiserver`, `kube-scheduler`, and `kubelet`.
You can enable (or disable) the feature by setting command line arguments:

```
--feature-gates="...,ReadWriteOncePod=true"
```

### Creating a PersistentVolumeClaim

In order to use the ReadWriteOncePod access mode for your PVs and PVCs, you will need to create a new PVC with the access mode:

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: single-writer-only
spec:
  accessModes:
  - ReadWriteOncePod # Allow only a single pod to access single-writer-only.
  resources:
    requests:
      storage: 1Gi
```

If your storage plugin supports [dynamic provisioning](https://kubernetes.io/docs/concepts/storage/dynamic-provisioning/), new PersistentVolumes will be created with the ReadWriteOncePod access mode applied.

#### Migrating existing PersistentVolumes

> If you have existing PersistentVolumes, they can be migrated to use ReadWriteOncePod.

Reviewer 2: Maybe also add something like "In this example, there is already a "cat-pictures-pvc" that is bound to a "cat-pictures-pv" PV."

As a first step, you need to edit your PersistentVolume's `spec.persistentVolumeReclaimPolicy` and set it to `Retain`.
This ensures your PersistentVolume will not be deleted when we delete the corresponding PersistentVolumeClaim.

> ```yaml

Reviewer 2: Is there an edit/patch command that can be used instead that can clearly show just the reclaim policy patch?

Author: Done

kind: PersistentVolume
apiVersion: v1
metadata:
  name: cat-pictures-pv
spec:
  accessModes:
    - ReadWriteOnce
  capacity:
    storage: 1Gi
  csi:
    driver: io.kubernetes.fake.driver
    volumeHandle: cat-pictures
  claimRef:
    name: cat-pictures-pvc
    namespace: default
  persistentVolumeReclaimPolicy: Retain # Ensure PV won't be deleted by accident.
  storageClassName: ""
```

Next you need to stop any workloads that are using the PersistentVolumeClaim bound to the PersistentVolume you want to migrate, and then delete the PersistentVolumeClaim:

```shell
kubectl scale --replicas=0 deployment cat-pictures-writer
kubectl delete pvc cat-pictures-pvc
```

> After that you need to replace the PersistentVolume's access modes with ReadWriteOncePod:

Reviewer 2: Same thing, can we have a patch command that only changes the accessMode?

In other cases maybe everyone is allowed to read data but not write it.
For highly sensitive data, maybe only one user is allowed to read and write data but nobody else.

In the world of Kubernetes, [access modes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes) are the way we model how storage is consumed.
These access modes are a part of the spec for PersistentVolumes (PVs) and PersistentVolumeClaims (PVCs).

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: shared-cache
spec:
  accessModes:
  - ReadWriteMany # Allow many pods to access shared-cache simultaneously.
  resources:
    requests:
      storage: 1Gi
  storageClassName: ""
```

In Kubernetes, we currently have three access modes for PVs and PVCs:

- ReadWriteOnce -- the volume can be mounted as read-write by a single node
- ReadOnlyMany -- the volume can be mounted read-only by many nodes
- ReadWriteMany -- the volume can be mounted as read-write by many nodes

These access modes are enforced by Kubernetes components like the `kube-controller-manager` and `kubelet` to ensure only certain pods are allowed to access a given PersistentVolume.

> This allows users to follow the principle of least privilege to mitigate the impact of bad actors.

Reviewer 2: I would remove this line because AccessModes is not really a security feature.

This could potentially be a major problem for some applications, especially if they require at most one writer for data safety guarantees.

With ReadWriteOncePod these issues go away.
Just set the access mode on your PVC, and Kubernetes will guarantee only a single pod has access.

## How do I use it?

The ReadWriteOncePod access mode is in alpha for Kubernetes 1.22 and is only supported for CSI volumes.
As a first step you need to enable the ReadWriteOncePod [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates) for `kube-apiserver`, `kube-scheduler`, and `kubelet`.
You can enable (or disable) the feature by setting command line arguments:

```
--feature-gates="...,ReadWriteOncePod=true"
```

### Creating a PersistentVolumeClaim

In order to use the ReadWriteOncePod access mode for your PVs and PVCs, you will need to create a new PVC with the access mode:

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: top-secret
spec:
  accessModes:
  - ReadWriteOncePod # Allow only a single pod to access top-secret.
  resources:
    requests:
      storage: 1Gi

>   storageClassName: ""

Reviewer 2: I think you'll want to have 2 different PVC examples: one for dynamic provisioning, and one for pre-existing volumes. 

For dynamic provisioning, the storageClassName should be empty (to use the default storageclass) or filled in.

Author: Removed the `storageClassName` field.

This dynamic provisioning example is different (at least the name is), were there other changes I should make to differentiate?

Author: Renamed this to `single-writer-only` (from `top-secret`) to clarify the intended use case for the access mode

  Warning  FailedScheduling  1s    default-scheduler  0/1 nodes are available: 1 node has pod using PersistentVolumeClaim with the same name and ReadWriteOncePod access mode.
```

### How is this different than the ReadWriteOnce access mode?

The ReadWriteOnce access mode restricts volume access to a single *node*, which means it is possible for multiple pods on the same node to read from and write to the same volume.
This could potentially be a major problem for some applications, especially if they require at most one writer for data safety guarantees.

With ReadWriteOncePod these issues go away.
Just set the access mode on your PVC, and Kubernetes will guarantee only a single pod has access.

## How do I use it?

The ReadWriteOncePod access mode is in alpha for Kubernetes 1.22 and is only supported for CSI volumes.
As a first step you need to enable the ReadWriteOncePod [feature gate](https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates) for `kube-apiserver`, `kube-scheduler`, and `kubelet`.
You can enable (or disable) the feature by setting command line arguments:

```
--feature-gates="...,ReadWriteOncePod=true"
```

> [here]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

Reviewer 6: is this line a left over?

Author: Removed, thank you!
